### ISSUE TITLE
Regarding the issue of type narrowing in conditional statements

### ISSUE THREAD

**Issue by Codfisher** [remeda/remeda#847]
English is not my first language, so I apologize in advance for any improper expressions or if my meaning is unclear. 

---

I saw the following code regarding conditionals in a document:

```ts
const nameOrId = 3 as string | number;
R.conditional(
  nameOrId,
  [R.isString, (name) => `Hello ${name}`],
  [R.isNumber, (id) => `Hello ID: ${id}`],
  R.conditional.defaultCase(
    (something) => `Hello something (${JSON.stringify(something)})`,
  ),
);
```

My understanding is that `isString` and `isNumber` can narrow the type of `nameOrId`.

Here is my code: [TS Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAcigFMQSATZQgKBpgE8wS4BxEmAEWRmQAVkUZAQC8cAN4040uI2YAuItQDcUmVV6KAjKoC+cAD41JM2UxKLCAI0KrTG5IoBMew8bXS5FogGNbHuAdFAGY9OmwAVwA7HxhgCCi4YFQAFXMAHjTmOBIADxgSKIp0di4efkFhAG1CL0IAXQA+AAovRSySAEoJANIYCKhE5rBKkEVS7l4BIRBOxRGZpPQAUXyhWPSA0wny6eEAGgkzBTgOuF0AxrgRK5NTGT6BxIXhADova5ExLzsZXTCaJEYnEEnAAOYcSbIYajcaQ3ajbp3aSPQaYV4+BIUYAgqLIAA2MJm+y2MiqpPuSw6zUI1E6JMplOa4kC5XO3RuPUZ3LgAHpeXAAOoACwYS1kwpYHwg2FZvDgUQg8C0AH4KTyHOrTLoGYz6rrpJ0aLogA)

I expected the `data` to be `1` after calling `isType('a')`, but that's not the case.

However, I noticed that the documentation mentions:

> Due to TypeScript's inability to infer the result of negating a type- predicate we can't refine the types used in subsequent cases based on previous conditions. Using a switch (true) statement or ternary operators is recommended for more precise type control when such type narrowing is needed.

In that case, why do `isString` and `isNumber` work in the example code?

Could it be that I am using it incorrectly? Or have I misunderstood how conditional types work?

I would greatly appreciate your advice on this.

**Comment by cjquines** [comment]
as a workaround, you can do

```ts
const isTypeA = isType('a');

R.conditional(param,
  [
    isTypeA,
    ({ data }) => {
      data
      // ^? (parameter) data: 1
    },
  ],
);
```

not sure if we can do something about this case, lemme investigate

**Comment by karl2015y** [comment]
I found that if 'isType' is not changed, and then modified to the following style, it can be narrowed successfully.  
So I guess maybe something was missing during the passing in the conditional.

[TS Playground](https://short.karllin.com/p/dKkKbjWmJzS4zkcE7ZWO)

```typescript
R.conditional(param,
        [
            (val) => isType('a')(val),
            ({ data }) => {
                // Why is the type of data not 1?
                data
            },
        ],
)
```

**Comment by eranhirsch** [comment]
From my previous attempts at fixing this, I noticed that `conditional` can't be easily fixed without `NoInfer`, which was added in TypeScript 5.4 (5.3?) so we won't be able to use that in the current major version of Remeda.

**Comment by eranhirsch** [comment]
After investigating this for the past couple of days and trying everything I could think of, I still couldn't resolve the underlying issue, even when using NoInfer.

There is something I don't fully understand regarding how TypeScript infers objects provided as parameters to a function and how they relate to type parameters for the function. TypeScript. To the best of my understanding, TypeScript would try to infer all elements of the object in parallel, even when they rely on one another, meaning that when it tries to infer the "then" function it infers it without looking at the shape of the "when" function, which means it takes the most broad function shape that satisfies it (e.g., `(data: T) => boolean`, and not `(data: T) => data is any)` which is a subset of it). Later, TypeScript would check if the resulting inference satisfies the whole of `conditional` and if so, it wouldn't try to infer a narrower type.

When wrapping the predicate with `($) => predicate($)`, we are effectively delaying this inference so that TypeScript has to deal with it first, before inferring `conditional`, and when it comes around to inferring `conditional`, that function is already narrowed.

At least this is what I came up with, trying every solution under the sun.

# TL;DR:

But! The silver lining here is that if we remove the tuples around each case and instead spread the cases as params, TypeScript infers them one by one, which allows it to narrow them:

```ts
// Instead of:
conditional(
  [when0, then0],
  [when1, then1],
  [when2, then2],
  [when3, then3],
  ...,
  fallback,
);

// We do:
conditional(
  when0,
  then0,
  when1,
  then1,
  when2,
  then2,
  ...,
  fallback,
);
```

This isn't a huge deal, because people are most likely using conditional with inline functions that are already multi-line, and the only difference for them would be the removal of the `[` and `]` around each case. For short one-liners, prettier would change the formatting to a longer and narrower column (as shown in the example above), but that's just a matter of formatting!

Because this is a breaking change, it'll have to wait for v3.

cc: @cjquines
